### ISSUE TITLE
Extend vault syntax to support multi-kv paths

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone#2353]
The current vault implementation assumes secrets are added using the `value` key:

```
vault wire secret/login/username value=xxx
vault wire secret/login/password value=xxx
```

However in many cases multiple secrets are added to the same path as multiple key-value pairs:

```
vault wire secret/login username=xxx password=xxx
```

To support paths with multiple key-value pairs, we will need to augment the current implementation so that one can specify which key to use. I have included a few different options below.

__disclaimer__ any options that we suggest need to be compatible with the docker-compose syntax, where the secrets section is of type `map[string]Secret`

__option 1:__ specify key in path, with delimiter

```
secrets:
  username:
    path: secret/login@username
```

__option 2:__ specify key as attribute

```
secrets:
  username:
    path: secret/login
    key: username
```

__option 3:__ enable a string OR object type. if just the path string is provide, we assume the key is value 

```
secrets:
  # supports: vault write secret/login/password value=xxx
  password:
    vault: secret/login/password
  # supports: vault write secret/login username=xxx
  username:
    vault:
      path: secret/login
      key: username
```

**Comment by gtaylor** [comment]
Option 2 is more verbose but also more obvious. I don't think I'd make the connection as to what's happening in option 3 without reading docs or asking in Slack.

**Comment by appleboy** [comment]
@bradrydzewski The vault feature only support enterprise version?

**Comment by mavimo** [comment]
I prefer option 1, this should be BC just considering string without `@` as `value` key, make YML file more readable (and compact).

**Comment by mavimo** [comment]
I suggest a poll using:

 - 😄 => 1
 - ❤️ => 2
 - 🎉 => 3

**Comment by jmccann** [comment]
I like option 1 and 2 the most from a usability perspective.

But wonder if something like option 3 is needed in case we have more secret plugins in the future ... like what if I wanted to use vault but also some other secrets tool?  Seems would need a way to identity which secret store to grab secrets from.  So not saying this is a plus for 3 necessarily, but maybe something that somehow needs to be considered for 1 and 2 also?

Also, I think all options should default to `value` key if nothing is explicitly set.  This would keep things backwards compatible (since specifying a key currently isn't required) and just make things easier since I think most people will use it in it's simplest form.

**Comment by jmccann** [comment]
Sorry, what I meant to say is people seem happy enough to use it for singular k/v pairs per secret.  

The `k` being the secret name and the `v` being the `value` as currently documented in http://docs.drone.io/vault-secrets/

I updated my above comment to try and clarify.

**Comment by snoby** [comment]
Hi @jmccann ,
If you look at the " First secret " part of the docs for vault you can see that writing multiple key values to a single secret is a valid use case.
https://www.vaultproject.io/intro/getting-started/first-secret.html
>You can even write multiple pieces of data, if you want:
```
$ vault write secret/hello value=world excited=yes
Success! Data written to: secret/hello
```
>vault write is a very powerful command. In addition to writing data directly from the command-line, it can read values and key pairs from STDIN as well as files. For more information, see the command documentation.

I will admit that I had not considered getting secrets from two different secret engines ( vault and the built in secret engine in drone) so option 3 is a valid answer in my workflow... Which has multiple key / values in a single secret.  

For reference we use this workflow to store certificates and the certificate key ( along with other bits associated with the certificate)

**Comment by snoby** [comment]
@bradrydzewski  Do we have a test version or some update on this?
thx

**Comment by bradrydzewski** [comment]
Thanks everyone for the feedback. I do have an initial implementation available in master and in the latest image in Dockerhub. The syntax and behavior is described below.

You can use the same syntax you use today (no breaking changes). This will use the `value` key by default to extract the secret. This syntax will be deprecated and removed from the documentation, but will continue to work pre and post 1.0 (so don't worry)

```
secrets:
  docker_username:
    path: secret/login/password
```

You can also use an extended syntax:

```
secrets:
  docker_username:
    driver: vault
    driver_opts:
      path: secret/login
      key: password
```

The latter implementation is fully compatible with the docker compose specification. If it seems a bit verbose it is because keeping parity with the specification is an important factor that drives how we implement YAML features.

You will also notice this change includes the driver field. This is going to be important going forward so that we can support various secret stores, including the amazon and google key stores, as well as other vault drivers and maybe even custom drivers.

```diff
secrets:
  docker_username:
+   driver: vault
```

**Comment by snoby** [comment]
Hi @bradrydzewski 
I found a bug in the new multikey implementation.  I have a test job that takes a base64 encoded text file and the length of the unencoded original length of the text file and I place those into a vault secret.  For example:
```
[:~] $ vault read secret/internal/CI/kubeconfig
Key                 Value
---                 -----
refresh_interval    768h
count               26476
value               YXBpVmVyc2lvbjogdjEKY2x1c3RlcnM6Ci0gY2x1...snippet
```
I have a bash script that pulls these values out of vault and makes them env values and then I run `drone exec` and everything runs and works.
```
[:~/work … cial/drone-test-vault] test_multikey_drone ± cat run_local_test_with_vault.sh
#!/bin/bash
set -x
#To run this test with `drone exec` you have to have already authenticated with vault.  You must also have jq installed

if [ -z "$KUBECONFIG64" ]; then
 KUBECONFIG64=$(vault read -format=json secret/internal/CI/kubeconfig | jq -r '.data.value')
 export KUBECONFIG64
fi
if [ -z "$count" ]; then
 COUNT=$(vault read -field=count secret/internal/CI/kubeconfig)
 export COUNT
fi

drone exec
```

However when I run the job on the actual drone server it fails...  The .drone.yaml looks like this:
```
    13	pipeline:
    14	  # The build step just obtains the kube config file and places it in the shared container directory
    15	  build:
    16	    image: alpine:latest
    17	    secrets: [KUBECONFIG64, COUNT]
    18	    volumes:
    19	      - kube_cache:/root/.kube
    20	    privileged: true
    21	    environment:
    22	      - KUBECONFIG=/root/.kube/config
    23	    commands:
    24	      - if [ -z "$COUNT" ]; then echo "Don't have the COUNT env variable defined"; exit 1;  fi
    25	      - if [ -z "$KUBECONFIG64" ]; then echo "Don't have the COUNT env variable defined"; exit 1; fi
    26	      - mkdir -p /root/.kube/
    27	      - echo -n "$KUBECONFIG64" | base64 -d > /root/.kube/config
    28	      - echo -n "$KUBECONFIG64" | base64 -d | wc -c > /tmp/count.txt
    29	      - echo "$COUNT" > /tmp/reference_count.txt
    30	      - DECODE_SIZE=$(cat /tmp/count.txt)
    31	      - if [ "$COUNT" != "$DECODE_SIZE" ]; then echo "Sizes don't match!";  echo "DECODE_SIZE=|$DECODE_SIZE| and COUNT=|$COUNT|"; exit 1; fi
    32	      - cat /tmp/reference_count.txt
    33	      - cat /tmp/count.txt
    34	      - diff -Nau /tmp/reference_count.txt /tmp/count.txt
```

It fails on line 31 because `count` is getting a line feed
here is the output:
```
+ if [ -z "$COUNT" ]; then echo "Don't have the COUNT env variable defined"; exit 1;  fi
+ if [ -z "$KUBECONFIG64" ]; then echo "Don't have the COUNT env variable defined"; exit 1; fi
+ mkdir -p /root/.kube/
+ echo -n "$KUBECONFIG64" | base64 -d > /root/.kube/config
+ echo -n "$KUBECONFIG64" | base64 -d | wc -c > /tmp/count.txt
+ echo "$COUNT" > /tmp/reference_count.txt
+ DECODE_SIZE=$(cat /tmp/count.txt)
+ if [ "$COUNT" != "$DECODE_SIZE" ]; then echo "Sizes don't match!";  echo "DECODE_SIZE=|$DECODE_SIZE| and COUNT=|$COUNT|"; exit 1; fi
Sizes don't match!
DECODE_SIZE=|26476| and COUNT=|26476
|
```
Notice that I put the pipe symbol '|' around the variables.  So for some reason when the drone server reads the value it's putting a line feed in where as when I do a vault read with -field or even -format=JSON there is no line feed.
Does this bug description make sense?

**Comment by snoby** [comment]
After a personal discussion with Brad i discovered that the way I was adding the count variable had a line feed in it, but when I was placing the count variable on to the ENV I was doing it with `echo -n` which was clipping the line feed.  I'm now removing the line feed when I add the secret:
`cat ~/.kube/allconfig | wc -c | tr -d ' '| tr  -d '\n'`

**Comment by bradrydzewski** [comment]
closing this issue. This feature is available in 0.8.5 image. Thanks everyone!
